src/emqtt.erl

```diff
@@ -1329,40 +1515,41 @@ do_ensure_retry_timer(Interval, State = #state{retry_timer = undefined})
 do_ensure_retry_timer(_Interval, State) ->
     State.
 
-retry_send(State = #state{inflight = Inflight}) ->
-    SortFun = fun({_, _, Ts1}, {_, _, Ts2}) -> Ts1 < Ts2 end,
-    Msgs = lists:sort(SortFun, maps:values(Inflight)),
-    retry_send(Msgs, os:timestamp(), State ).
-
-retry_send([], _Now, State) ->
-    {keep_state, ensure_retry_timer(State)};
-retry_send([{Type, Msg, Ts} | Msgs], Now, State = #state{retry_interval = Interval}) ->
-    Diff = timer:now_diff(Now, Ts) div 1000, %% micro -> ms
-    case (Diff >= Interval) of
-        true  -> case retry_send(Type, Msg, Now, State) of
-                     {ok, NewState} -> retry_send(Msgs, Now, NewState);
-                     {error, Error} -> {stop, Error}
-                 end;
-        false -> {keep_state, do_ensure_retry_timer(Interval - Diff, State)}
+sent_at(?INFLIGHT_PUBLISH(_, SentAt, _, _)) ->
+    SentAt;
+sent_at(?INFLIGHT_PUBREL(_, SentAt, _)) ->
+    SentAt.
+
+retry_send(State = #state{retry_interval = Intv, inflight = Inflight}) ->
+    try
+        Now = now_ts(),
+        Pred = fun(_, InflightReq) ->
+                       (sent_at(InflightReq) + timer:seconds(Intv)) > Now end,
+        NInflight = lists:foldl(
```

Comment:
nit: prefer the old style recursive loop over `lists:foldl`

Reply:
I've made some changes, but I feel as if it's still not concise and easy to read.